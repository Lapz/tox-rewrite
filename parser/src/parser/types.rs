use crate::parser::{Parser, Precedence, Restrictions};
use crate::SyntaxKind::*;
use syntax::T;

impl<'a> Parser<'a> {
    pub(crate) fn parse_type(&mut self) {
        match self.current() {
            IDENT | T![self] => self.parse_ident_type(),
            T!["["] => self.parse_array_type(),
            T!["("] => self.parse_paren_type(),
            T![fn] => self.parse_fn_type(),
            _ => self.error(
                "Expected a type parameter",
                format!(
                    "Expected a type parameter but instead found `{}`",
                    self.current_string()
                ),
            ),
        };
    }

    fn parse_ident_type(&mut self) {
        self.start_node(IDENT_TYPE);

        if self.matches(vec![IDENT, T![self]]) {
            self.bump();
        } else {
            self.error(
                "Expected an identifier or `void`",
                format!(
                    "Expected an identifier or `void` but instead found `{}`",
                    self.current_string()
                ),
            )
        }

        if self.at(T![<]) {
            self.parse_type_args();
        }

        self.finish_node();
    }

    fn parse_array_type(&mut self) {
        self.start_node(ARRAY_TYPE);
        self.bump(); //Eat `[`

        self.parse_type();

        if self.at(T![;]) {
            self.bump();
            self.parse_expression(Precedence::Primary, Restrictions::default())
        }

        self.expect(T!["]"]);

        self.finish_node();
    }

    fn parse_paren_type(&mut self) {
        self.start_node(PAREN_TYPE);
        self.bump(); //Eat `(`

        while !self.at(EOF) && !self.at(T![")"]) {
            self.parse_type();
            if !self.at(T![")"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![")"]);

        self.finish_node();
    }

    fn parse_fn_type(&mut self) {
        self.start_node(FN_TYPE);

        self.expect(T![fn]);

        self.expect(T!["("]);

        while !self.at(EOF) && !self.at(T![")"]) {
            self.parse_type();
            if !self.at(T![")"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![")"]);

        if self.at(T![->]) {
            self.start_node(RET_TYPE);
            self.bump();
            self.parse_type();
            self.finish_node();
        }

        self.finish_node();
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_fn_type,"fn main(_:fn(i32,i32) -> i32) {}"}
    test_parser! {parse_fn_tuple_type,"fn main(_:fn((i32,i32)) -> i32) {}"}
    test_parser! {parse_ident_type,"fn main(x:i32) {}"}
    test_parser! {parse_array_type,"fn main(x:[i32]) {}"}
    test_parser! {parse_array_tuple_type,"fn main(x:[(i32,i32)]) {}"}
    test_parser! {parse_tuple_type,"fn main(x:(i32,i32)) {}"}
}
