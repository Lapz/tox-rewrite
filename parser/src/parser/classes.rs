use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_class(&mut self, checkpoint: rowan::Checkpoint) {
        self.start_node_at(checkpoint, CLASS_DEF);

        self.expect(CLASS_KW);

        self.ident();

        if self.at(T![<]) {
            self.parse_type_params(false);
        }

        self.parse_class_body();

        self.finish_node()
    }

    fn parse_class_body(&mut self) {
        self.expect(T!["{"]);

        while !self.at(EOF) && !self.at(T!["}"]) {
            let has_visibility = self.has_visibility();

            if has_visibility {
                let checkpoint = self.checkpoint();
                self.parse_visibility();

                match self.current() {
                    IDENT => self.parse_named_field(),
                    T![fn] => self.parse_function(checkpoint),

                    _ => self.error(
                        "Expected an identifier | `pub` | `fn` ",
                        format!(
                            "Expected an identifier | `pub` | `fn`  but instead found `{}`",
                            self.current_string()
                        ),
                    ),
                }
            } else {
                let checkpoint = self.checkpoint();
                match self.current() {
                    IDENT => self.parse_named_field(),
                    T![fn] => self.parse_function(checkpoint),

                    _ => self.error(
                        "Expected an identifier | `pub` | `fn` ",
                        format!(
                            "Expected an identifier | `pub` | `fn`  but instead found `{}`",
                            self.current_string()
                        ),
                    ),
                }
            }
        }

        self.expect(T!["}"]);
    }

    fn parse_named_field(&mut self) {
        self.start_node(NAMED_FIELD_DEF);
        self.ident();
        self.expect(T![:]);
        self.parse_type();
        self.expect(T![;]);
        self.finish_node();
    }
}

#[cfg(test)]
mod tests {

    use crate::utils::parse;
    use syntax::{ClassDefOwner, NamedFieldsOwner};
    #[test]
    fn test_class_fields() {
        let source_file = parse("class Person { name:String; surname:String;}");

        let class = source_file.classes().next().unwrap();

        assert_eq!(class.fields().count(), 2)
    }
    test_parser! {parse_empty_class,"class Foo {}"}
    test_parser! {parse_exported_class,"export class Foo {}"}
    test_parser! {parse_class_generic,"class Result<T,E> {}"}
    test_parser! {parse_class_fields,"class Person { name:String; surname:String;}"}
    test_parser! {parse_class_fields_methods,"class Person { name:String; surname:String; fn hello(self) {}}"}
    test_parser! {parse_class_methods,"class Person { name:String; surname:String; fn new() -> Person {}}"}
}
