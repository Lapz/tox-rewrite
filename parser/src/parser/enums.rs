use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_enum(&mut self, checkpoint: rowan::Checkpoint) {
        self.start_node_at(checkpoint, ENUM_DEF);

        self.expect(T![enum]);

        self.ident();

        if self.at(T![<]) {
            self.parse_type_params(false);
        }

        self.parse_enum_variants();

        self.finish_node()
    }

    fn parse_enum_variants(&mut self) {
        self.start_node(ENUM_VARIANT_LIST);

        self.expect(T!["{"]);

        while !self.at(EOF) && !self.at(T!["}"]) {
            self.parse_enum_variant();
            if !self.at(T!["}"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T!["}"]);

        self.finish_node();
    }

    fn parse_enum_variant(&mut self) {
        self.start_node(ENUM_VARIANT);

        self.ident();

        if self.at(T!["("]) {
            self.parse_enum_variant_types()
        }

        self.finish_node();
    }

    fn parse_enum_variant_types(&mut self) {
        self.bump(); // Eat the "("
        while !self.at(EOF) && !self.at(T![")"]) {
            self.parse_type();

            if !self.at(T![")"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![")"]);
    }
}

#[cfg(test)]
mod tests {
    test_parser! {parse_empty_enum,"enum Foo {}"}
    test_parser! {parse_exported_enum,"export enum Foo {}"}
    test_parser! {parse_enum_with_variants,"enum Foo {A,B,C}"}
    test_parser! {parse_enum_variants_trailing_comma,"enum Foo {A,B,C,}"}
    test_parser! {parse_enum_variants_types,"enum Foo { A([i32]),B(i32),C((i32,i32)),}"}
    test_parser! {parse_enum_generic,"enum Result<T,E> { Ok(T),Err(E)}"}
}
