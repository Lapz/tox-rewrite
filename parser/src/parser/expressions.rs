use crate::parser::{
    pratt::{Precedence, Rule, RuleToken},
    Parser, Restrictions,
};

use syntax::T;

mod binary;
mod block;
mod break_expr;
mod call_expr;
mod closure_expr;
mod continue_expr;
mod do_expr;

mod field_expr;
mod for_expr;
mod grouping;
mod ident;
mod if_expr;
mod index_expr;
mod let_expr;
mod literal;
mod match_expr;
mod record_expr;
mod return_expr;
mod unary;
mod while_expr;

pub use binary::BinaryParselet;
pub use call_expr::CallParselet;
pub use closure_expr::ClosureParselet;
pub use field_expr::FieldParselet;
pub use grouping::GroupingParselet;
pub use ident::IdentParselet;
pub use index_expr::IndexParselet;
pub use literal::LiteralParselet;
pub use record_expr::RecordParselet;
pub use unary::UnaryParselet;

impl<'a> Parser<'a> {
    pub(crate) fn parse_expression(&mut self, precedence: Precedence, restrictions: Restrictions) {
        let check_point = self.builder.checkpoint();

        let token = self.current();

        let mut rule = token.rule();

        let parser = self.prefix.get(&rule);

        let parser = if let Some(parser) = parser {
            parser
        } else {
            self.error(
                "Expected an expression",
                format!(
                    "Expected an expression but instead found `{}`",
                    self.current_string()
                ),
            );
            return;
        };

        parser.parse(self);

        while precedence <= self.precedence() {
            {
                let token = self.current();
                rule = token.rule();
            }

            let parser = self.infix.get(&rule);

            let parser = if parser.is_some() {
                parser.unwrap()
            } else {
                self.error(
                    "Expected an expression",
                    format!(
                        "Expected an expression but instead found `{}`",
                        self.current_string()
                    ),
                );
                break;
            };

            if restrictions.forbid_record && rule == RuleToken::LBrace {
                break;
            }
            parser.parse(self, check_point);
        }

        if precedence == Precedence::None {
            self.recover_until(T![;]);
        }
    }
}
