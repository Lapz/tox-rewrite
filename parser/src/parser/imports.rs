use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_import(&mut self) {
        self.start_node(IMPORT_DEF);

        self.expect(T![import]);

        while !self.at(EOF) && !self.at(T![;]) {
            self.parse_import_segment();

            if !self.at(T![;]) && !self.expected(T![::]) {
                break;
            }
        }

        self.expect(T![;]);

        self.finish_node();
    }

    pub(crate) fn parse_import_segment(&mut self) {
        if self.at(T!["{"]) {
            self.start_node(IMPORT_LIST);

            self.bump();

            while !self.at(EOF) && !self.at(T!["}"]) {
                self.parse_import_segment();

                if !self.at(T!["}"]) && !self.expects(vec![T![,], T![::]]) {
                    break;
                }
            }

            self.expect(T!["}"]);
        } else if !self.at(T![;]) {
            self.start_node(IMPORT_SEGMENT);
            self.ident();
        }

        self.finish_node();
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_empty_import,"import foo;"}
    test_parser! {parse_import_with_multiple_segments,"import foo::bar::baz;"}
    test_parser! {parse_import_with_brace,"import foo::{bar,baz};"}
    test_parser! {parse_import_with_brace_nested_segments,"import foo::{bar::baz,fubar};"}
}
