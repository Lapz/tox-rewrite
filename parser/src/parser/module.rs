use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_module(&mut self) {
        self.start_node(MOD_DEF);

        self.expect(T![mod]);

        self.ident();

        self.expect(T![;]);

        self.finish_node();
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_single_mod,"mod foo;"}
    test_parser! {parse_many_mods,"mod bar; mod foo;"}
}
