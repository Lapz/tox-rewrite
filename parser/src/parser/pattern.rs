use crate::parser::{Parser, Precedence, Restrictions};
use syntax::T;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_pattern(&mut self, allow_literal: bool) {
        match self.current() {
            T!["("] => self.parse_tuple_pattern(allow_literal),
            IDENT => self.parse_binding_pattern(),
            T![self] => self.bump(),
            T![_] => self.parse_placeholder_pattern(),
            e => {
                if allow_literal {
                    self.start_node(LITERAL_PAT);
                    self.parse_expression(Precedence::Primary, Restrictions::default());
                    self.finish_node()
                } else {
                    self.error(
                        "Expected a literal pattern",
                        format!(
                            "Expected a literal pattern instead found `{}` {:?}",
                            self.current_string(),
                            e
                        ),
                    )
                }
            }
        }
    }

    fn parse_tuple_pattern(&mut self, allow_literal: bool) {
        self.start_node(TUPLE_PAT);

        self.bump(); // Eat the `(`

        while !self.at(EOF) && !self.at(T![")"]) {
            self.parse_pattern(allow_literal);
            if !self.at(T![")"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![")"]);
        self.finish_node();
    }

    fn parse_placeholder_pattern(&mut self) {
        self.start_node(PLACEHOLDER_PAT);
        self.expect(T![_]);
        self.finish_node();
    }

    fn parse_binding_pattern(&mut self) {
        self.start_node(BIND_PAT);
        self.ident();
        self.finish_node();
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_placeholder_pattern,"fn main(_:i32) {}"}

    test_parser! {parse_tuple_pattern,"fn main((x,y):i32) {}"}

    test_parser! {parse_binding_pattern,"fn main(x:i32) {}"}

    test_parser! {parse_nested_tuple_pattern,"fn main((x,(y,_)):i32) {}"}
}
