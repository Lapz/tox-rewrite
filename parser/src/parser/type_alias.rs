use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_type_alias(&mut self, checkpoint: rowan::Checkpoint) {
        self.start_node_at(checkpoint, TYPE_ALIAS_DEF);

        self.expect(T![type]);

        self.ident();

        if self.at(L_ANGLE) {
            self.parse_type_params(false);
        }

        self.expect(EQ);

        self.parse_type();

        self.expect(SEMI);

        self.finish_node()
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_type_alias,"type Foo = i32;"}
    test_parser! {parse_type_alias_params,"type ParseResult<T> = Result<T,void>;"}
}
