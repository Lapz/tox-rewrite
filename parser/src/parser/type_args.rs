use crate::parser::Parser;
use crate::SyntaxKind::*;
use syntax::T;

impl<'a> Parser<'a> {
    pub(crate) fn parse_type_args(&mut self) {
        self.start_node(TYPE_ARG_LIST);

        self.bump();

        while !self.at(EOF) && !self.at(T![>]) {
            self.parse_type();

            if !self.at(T![>]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![>]);

        self.finish_node();
    }
}
