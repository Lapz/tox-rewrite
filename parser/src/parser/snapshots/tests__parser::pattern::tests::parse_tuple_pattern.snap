---
created: "2020-05-06T10:43:13.320768100Z"
creator: insta@0.8.2
source: parser/src/parser/pattern.rs
expression: "crate::utils::dump_debug(&parser_output)"
---
SOURCE_FILE@[0; 21)
  FN_DEF@[0; 21)
    FN_KW@[0; 2) "fn"
    WHITESPACE@[2; 3) " "
    NAME@[3; 7)
      IDENT@[3; 7) "main"
    PARAM_LIST@[7; 18)
      L_PAREN@[7; 8) "("
      PARAM@[8; 17)
        TUPLE_PAT@[8; 13)
          L_PAREN@[8; 9) "("
          BIND_PAT@[9; 10)
            NAME@[9; 10)
              IDENT@[9; 10) "x"
          COMMA@[10; 11) ","
          BIND_PAT@[11; 12)
            NAME@[11; 12)
              IDENT@[11; 12) "y"
          R_PAREN@[12; 13) ")"
        COLON@[13; 14) ":"
        IDENT_TYPE@[14; 17)
          IDENT@[14; 17) "i32"
      R_PAREN@[17; 18) ")"
    WHITESPACE@[18; 19) " "
    BLOCK_EXPR@[19; 21)
      BLOCK@[19; 21)
        L_CURLY@[19; 20) "{"
        R_CURLY@[20; 21) "}"

